### PR TITLE
Support Module aliasing in SwiftPM

### PR DESCRIPTION
**Pull request by elsh**
Resolves rdar://87583788

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test

**Comment by tomerd** [comment]
@swift-ci please smoke test

**Comment by elsh** [comment]
https://github.com/apple/swift-llbuild/pull/758
@swift-ci smoke test

**Comment by elsh** [comment]
apple/swift-llbuild#758
@swift-ci smoke test

**Comment by elsh** [comment]
apple/swift-llbuild#758
@swift-ci Please Build Toolchain macOS Platform

**Comment by tomerd** [comment]
@elsh in order to build a toolchain, you need to create a PR on the swift repositories and then mention all the related PR (like this one and the llbuild one) for cross repo building. iow, the SwiftPM repo is not setup to create toolchains, you have to create an empty PR on the swift toolchain to do that.

File: Sources/PackageDescription/Target.swift [hunk 1]
```diff
@@ -981,15 +986,25 @@ extension Target.Dependency {
...
     ) -> Target.Dependency {
-        return .productItem(name: name, package: package, condition: condition)
+        return .productItem(name: name, moduleAliases: nil, package: package, condition: condition)
     }
-    
+
+    @available(_PackageDescription, introduced: 999.0)
+    public static func product(
+      name: String,
+      moduleAliases: [String: String]? = nil,
```
Comment (tomerd) [inline Sources/PackageDescription/Target.swift:1001]: I think the right order is: `name`, `package`, `moduleAliases`, `condition`: defaulted argument come after non-defaulted ones as a rule of thumb

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 2]
```diff
@@ -222,15 +222,56 @@ private func createResolvedPackages(
...
+                }
+            }
+        }
+    }
+  
     // Scan and validate the dependencies
     for packageBuilder in packageBuilders {
         let package = packageBuilder.package
-
+
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:267]: nit: remove dead space

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 3]
```diff
@@ -222,15 +222,56 @@ private func createResolvedPackages(
...
+        if case let .product(prodRef, _) = d {
+            if let moduleAliasesForDep = prodRef.moduleAliases, let depPkg = prodRef.package {
+                for (key, val) in moduleAliasesForDep {
+                    moduleAliasMap[key] = (val, depPkg)
+                }
+            }
+        }
+    }}}
+  
+    func setModuleAliasesForTargets(in pkg: Package) {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:237]: should this be a private extension of Package instead of a nested function that takes package?

File: Sources/PackageModel/Manifest/TargetDescription.swift [hunk 4]
```diff
@@ -24,15 +24,15 @@ public struct TargetDescription: Equatable, Codable {
     /// Represents a target's dependency on another entity.
     public enum Dependency: Equatable {
         case target(name: String, condition: PackageConditionDescription?)
-        case product(name: String, package: String?, condition: PackageConditionDescription?)
+        case product(name: String, moduleAliases: [String: String]? = nil, package: String?, condition: PackageConditionDescription?)
```
Comment (tomerd) [inline Sources/PackageModel/Manifest/TargetDescription.swift:27]: I think the order should be name, package, moduleAliases, condition to be consistent with the comment above

File: Sources/PackageModel/ManifestSourceGeneration.swift [hunk 5]
```diff
@@ -319,8 +319,11 @@ fileprivate extension SourceCodeFragment {
             }
             self.init(enum: "target", subnodes: params)
             
-        case .product(name: let name, package: let packageName, condition: let condition):
+        case .product(name: let name, moduleAliases: let aliases, package: let packageName, condition: let condition):
```
Comment (tomerd) [inline Sources/PackageModel/ManifestSourceGeneration.swift:322]: should we use the opportunity to clean up the syntax here?

eg `.product(let name, let aliases, let packageName, let condition)`

File: Sources/PackageModel/Target.swift [hunk 6]
```diff
@@ -37,11 +37,15 @@ public class Target: PolymorphicCodableProtocol {
...
 
+        /// Module aliases for targets of this product dependency.
+        public let moduleAliases: [String: String]?
+
         /// The name of the package containing the product.
         public let package: String?
 
         /// Creates a product reference instance.
-        public init(name: String, package: String?) {
+        public init(name: String, moduleAliases: [String: String]? = nil, package: String?) {
```
Comment (tomerd) [inline Sources/PackageModel/Target.swift:47]: same order as mentioned above

File: Sources/PackageModel/Target.swift [hunk 7]
```diff
@@ -104,16 +108,34 @@ public class Target: PolymorphicCodableProtocol {
...
     /// name) name in many cases, instead use c99name if you need uniqueness.
-    public let name: String
+    public private(set) var name: String
+
+    /// Module aliases needed to build this target.
+    public private(set) var moduleAliases: [String: String]?
 
+    /// Set module aliases needed to build this target.
+    /// shouldRename True if the target itself should be renamed
+    public func setModuleAliases(name: String, alias: String, shouldRename: Bool = false) {
```
Comment (tomerd) [inline Sources/PackageModel/Target.swift:118]: should this be a setter instead of a func?

Comment (tomerd) [reply Sources/PackageModel/Target.swift:118]: can you explain a bit when `shouldRename` is used, it's interesting that its the caller responsibility to decide that?

File: Tests/BuildTests/BuildPlanTests.swift [hunk 8]
```diff
@@ -101,6 +101,62 @@ final class BuildPlanTests: XCTestCase {
...
+        observabilityScope: observability.topScope
+      ))
+
+      result.checkProductsCount(1)
+      result.checkTargetsCount(4)
+
+      XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "OtherLogging" && $0.target.moduleAliases?["Logging"] == "OtherLogging" })
+      XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases?["Logging"] == "OtherLogging" })
+      XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Logging" && $0.target.moduleAliases == nil })
+    }
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:158]: do wen need to add some edge case and collision / negative use case tests?

Comment (elsh) [reply Tests/BuildTests/BuildPlanTests.swift:158]: yes will add

File: Sources/PackageModel/Target.swift (hunk 7 above)
Comment (elsh) [reply Sources/PackageModel/Target.swift:118]: It's set to true if the target name is the one that needs to be aliased as opposed to its dependency name; will look into autodetecting it / removing the param.

File: Sources/PackageModel/ManifestSourceGeneration.swift (hunk 5 above)
Comment (elsh) [reply Sources/PackageModel/ManifestSourceGeneration.swift:322]: sg

File: Sources/PackageDescription/Target.swift (hunk 1 above)
Comment (elsh) [reply Sources/PackageDescription/Target.swift:1001]: 👍

File: Sources/LLBuildManifest/Tools.swift [hunk 9]
```diff
@@ -221,6 +221,7 @@ public struct SwiftCompilerTool: ToolProtocol {
 
     public var executable: AbsolutePath
     public var moduleName: String
+    public var moduleAliases: [String: String]?
```
Comment (abertelrud) [inline Sources/LLBuildManifest/Tools.swift:224]: Would it make sense to have this as a dictionary that's usually empty or does an optional dictionary provide a semantic value?

Comment (abertelrud) [reply Sources/LLBuildManifest/Tools.swift:224]: This comment applies throughout the code, this line was just the first place I thought of it.

**Comment by elsh** [comment]
@swift-ci test

File: Sources/LLBuildManifest/Tools.swift (hunk 9 above)
Comment (elsh) [reply Sources/LLBuildManifest/Tools.swift:224]: Either is ok I think. In the `func write(to:)`  it checks for nil and add the attribute only if non-nil.

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

File: Sources/PackageDescription/Target.swift [hunk 10]
```diff
@@ -981,15 +986,25 @@ extension Target.Dependency {
...
         package: String,
         condition: TargetDependencyCondition? = nil
     ) -> Target.Dependency {
-        return .productItem(name: name, package: package, condition: condition)
+        return .productItem(name: name, package: package, moduleAliases: nil, condition: condition)
     }
-    
+
+    @available(_PackageDescription, introduced: 999.0)
+    public static func product(
```
Comment (tomerd) [inline Sources/PackageDescription/Target.swift:999]: needs API doc

File: Sources/PackageLoading/PackageBuilder.swift [hunk 11]
```diff
@@ -718,7 +717,7 @@ public final class PackageBuilder {
                     }
                 }
             } ?? []
-
+
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:720]: kill redundant space (you can setup Xcode to do this automatically)

File: Sources/PackageLoading/PackageBuilder.swift [hunk 12]
```diff
@@ -661,7 +661,7 @@ public final class PackageBuilder {
         var targets = [String: Target]()
         // If a direcotry is empty, we don't create a target object for them.
         var emptyModules = Set<String>()
-
+
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:664]: kill redundant space (you can setup Xcode to do this automatically)

File: Sources/PackageLoading/PackageBuilder.swift [hunk 13]
```diff
@@ -590,7 +590,7 @@ public final class PackageBuilder {
             let path = try findPath(for: target)
             return PotentialModule(name: target.name, path: path, type: target.type)
         })
-
+
```
Comment (tomerd) [inline Sources/PackageLoading/PackageBuilder.swift:593]: kill redundant space (you can setup Xcode to do this automatically)

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 14]
```diff
@@ -239,14 +265,21 @@ private func createResolvedPackages(
...
     }
 
+    // Gather all module aliases specified for dependencies from each package
+    let pkgToAliasesMap = gatherModuleAliases(from: packageBuilders)
+
     // Scan and validate the dependencies
     for packageBuilder in packageBuilders {
         let package = packageBuilder.package
-
+
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:274]: kill redundant space (you can setup Xcode to do this automatically)

File: Tests/BuildTests/BuildPlanTests.swift [hunk 15]
```diff
@@ -1636,6 +1636,149 @@ final class BuildPlanTests: XCTestCase {
             planResult.checkTargetsCount(5)
         }
     }
+    
+    func testModuleAliasingDirectDeps() throws {
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:1640]: format... you can use Xcode control+I to automatically fix formatting

File: Tests/BuildTests/BuildPlanTests.swift [hunk 16]
```diff
@@ -1636,6 +1636,149 @@ final class BuildPlanTests: XCTestCase {
...
+        
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "OtherLogging" && $0.target.moduleAliases?["Logging"] == "OtherLogging" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Utils" && $0.target.moduleAliases?["Logging"] == "OtherLogging" })
+        XCTAssertTrue(result.targetMap.values.contains { $0.target.name == "Logging" && $0.target.moduleAliases == nil })
+        #if os(macOS)
+            let otherLoggingArgs = try result.target(for: "OtherLogging").swiftTarget().compileArguments()
+            let loggingArgs = try result.target(for: "Logging").swiftTarget().compileArguments()
+            XCTAssertMatch(otherLoggingArgs, [.anySequence, "-emit-objc-header", "-emit-objc-header-path", "/path/to/build/debug/OtherLogging.build/OtherLogging-Swift.h", .anySequence])
+            XCTAssertMatch(loggingArgs, [.anySequence, "-emit-objc-header", "-emit-objc-header-path", "/path/to/build/debug/Logging.build/Logging-Swift.h", .anySequence])
+        #endif
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:1780]: is there an equivalent compileArguments  check we need to do in non-macOS cases?

**Review by tomerd — COMMENTED** [review]
very nice. couple of small comments

**Comment by tomerd** [comment]
probably worth adding at least one fixtures in the functional tests area for a end-to-end test of the functionality

File: Sources/PackageDescription/Target.swift (hunk 10 above)
Comment (elsh) [reply Sources/PackageDescription/Target.swift:999]: Should we call this new parameter `targetAliases` instead of `moduleAliases` btw? Any opinions on this @neonichu @abertelrud @tomerd ?

Comment (abertelrud) [reply Sources/PackageDescription/Target.swift:999]: No, I think `moduleAliases` is the best, since not all targets produce modules and the alias here is really for the module name, not the target name.  We should actually have a parameter to optionally control the module name here (right now it's just always mangled from the target name) but that can be added separately.

File: Tests/BuildTests/BuildPlanTests.swift (hunk 16 above)
Comment (elsh) [reply Tests/BuildTests/BuildPlanTests.swift:1789]: Added negative tests similar to other tests checking this flag.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
> probably worth adding at least one fixtures in the functional tests area for a end-to-end test of the functionality

Planning to add that in a separate PR; lmk if you think it should accompany this PR.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
i think at least one end-to-end (fixture) test is in place

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

File: Fixtures/Miscellaneous/ModuleAliasing/Simple/appPkg/Package.swift [hunk 17]
```diff
@@ -2,23 +2,23 @@
 import PackageDescription
```
Comment (tomerd) [inline Fixtures/Miscellaneous/ModuleAliasing/Simple/appPkg/Package.swift:2]: I think you need to update the path on disk (in git) to also be AppPkg and  GamePkg, they look to be still camel case which would cause issues on case sensitive file systems. git can unfortunately be a pain about that

File: Fixtures/Miscellaneous/ModuleAliasing/Simple/gamePkg/Package.swift [hunk 18]
```diff
@@ -1,8 +1,8 @@
-// swift-tools-version:4.2
```
Comment (tomerd) [inline Fixtures/Miscellaneous/ModuleAliasing/Simple/gamePkg/Package.swift]: https://github.com/apple/swift-package-manager/pull/4023/commits/499356211f5101df9563c1876cb96c0c978a4488#r802030268

**Review by tomerd — APPROVED** [review]
looks good once fixture path are fixed

File: Fixtures/Miscellaneous/ModuleAliasing/Simple/appPkg/Package.swift (hunk 17 above)
Comment (elsh) [reply Fixtures/Miscellaneous/ModuleAliasing/Simple/appPkg/Package.swift:2]: strange, I renamed the paths but somehow the change is not showing up, will redo.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
🥳
